src/sched/basic.c

```diff
@@ -112,8 +112,7 @@ static void sched_run(ABT_sched sched)
         for (i = 0; i < num_pools; i++) {
             ABT_pool pool = pools[i];
             ABTI_pool *p_pool = ABTI_pool_get_ptr(pool);
-            size_t size = ABTI_pool_get_size(p_pool);
-            if (size > 0) {
+            if (num_pools == 1 || ABTI_pool_get_size(p_pool) > 0 ) {
```

Comment:
How about we completely get rid of this branch? `get_size > 0` and the beginning of `pop` do the same thing; they check that `num_units > 0`. So might as well save on the redundant branch

Reply:
The problem is that the new pool added by this PR (called blocking_fifo in this PR, but possibly to be renamed re: offline discussion) can block on the pop() function.  If you have more than one pool associated with a scheduler then the other pool could be starved if you are using that pool.

I could rework it so that the existing schedulers have the optimization that you describe, and then add a new scheduler that is aware of the blocking behavior and responds accordingly.  We already have to start a new scheduler to swap pools anyway, so it could just as well be a special scheduler intended for use with the blocking_fifo pool.

That's a little bit of code duplication, but it would let us cut down the code path for the default schedulers.